src/mapreduce.jl

```diff
@@ -285,3 +285,43 @@ end
         @inbounds return similar_type(a, T, Size($Snew))(tuple($(exprs...)))
     end
 end
+
+struct _InitialValue end
+
+_maybeval(dims::Integer) = Val(Int(dims))
+_maybeval(dims) = dims
+_valof(::Val{D}) where D = D
+
+@inline Base.accumulate(op::F, a::StaticVector; dims = :, init = _InitialValue()) where {F} =
+    _accumulate(op, a, _maybeval(dims), init)
+
+@inline Base.accumulate(op::F, a::StaticArray; dims, init = _InitialValue()) where {F} =
+    _accumulate(op, a, _maybeval(dims), init)
+
+@inline function _accumulate(op::F, a::StaticArray, dims::Union{Val,Colon}, init) where {F}
+    # Adjoin the initial value to `op`:
+    rf(x, y) = x isa _InitialValue ? y : op(x, y)
+
+    # StaticArrays' `reduce` is `foldl`:
```

Comment:
> Do you think `StaticArrays` should rename our internal implementation of `reduce` to `foldl` for clarity?

Yeah, I'd prefer that; i.e., define `foldl` and then call it via `reduce`.

(Also, maybe having "non-sequential" version of `reduce` (e.g., `(a+b)+(c+d)`) actually makes sense?  I wonder if it helps compiler/hardware using instruction-level parallelism.)

> We've had various problems with named tuples destroying performance that perhaps we should just strip them ASAP for now and never use them in the internals.

I see that `NamedTuple` in `StaticArrays` is used for denoting if `init` is given or not at type-level.  If you are OK with adding the special `_InitialValue` type as done in this PR, I don't think we need to use `NamedTuple` anymore.

Reply:
> We've had various problems with named tuples destroying performance that perhaps we should just strip them ASAP for now and never use them in the internals.

Can this rewrite be done in another PR?  That's not a tiny one as it has to touch all the code implementing `mapreduce` and `reduce`.

Reply:
> Also, maybe having "non-sequential" version of `reduce` (e.g., `(a+b)+(c+d)`) actually makes sense? I wonder if it helps compiler/hardware using instruction-level parallelism.)

Agreed! Actually I had wondered the same thing and I wouldn't be surprised if it was much faster.

Something similar could be said of `cumsum` actually: it would be rather interesting to implement it as a work-efficient parallel prefix sum and see whether the improved ILP would compensate for the extra operations.

> I see that `NamedTuple` in `StaticArrays` is used for denoting if `init` is given or not at type-level. If you are OK with adding the special `_InitialValue` type as done in this PR, I don't think we need to use `NamedTuple` anymore.

Yes if we're after a special type to specify "init not given" I think it would be fine to use `_InitialValue`. Though perhaps we should rename it to `_EmptyInit` or something?

> Can this rewrite be done in another PR? That's not a tiny one as it has to touch all the code implementing `mapreduce` and `reduce`.

Sure, good plan.

Reply:
> Yes if we're after a special type to specify "init not given" I think it would be fine to use `_InitialValue`. Though perhaps we should rename it to `_EmptyInit` or something?

I've already shoe-horned [`_InitialValue` in `base/reduce.jl`](https://github.com/JuliaLang/julia/blob/3058112b0e09acae3305bd2f102657c8b059f7a2/base/reduce.jl#L64) :smile: so I thought it might be better to use the same name.

Reply:
Ok, let's go with that.